internal/ssh/limitlistener/listener_test.go

```diff
@@ -0,0 +1,409 @@
+	go func() {
+		defer wg.Done()
+
+		accepted := 0
+		for {
+			c, err := l.Accept()
+			if err != nil {
+				break
+			}
+			accepted++
+			if accepted == max {
+				close(saturated)
+			}
+			_, _ = io.WriteString(c, msg)
+
+			// Leave c open until the listener is closed.
+			defer c.Close()
+		}
+		t.Logf("with limit %d, accepted %d simultaneous connections", max, accepted)
+		// The listener accounts open connections based on Listener-side Close
+		// calls, so even if the client hangs up early (for example, because it
+		// was a random dial from another process instead of from this test), we
+		// should not end up accepting more connections than expected.
+		if accepted != max {
+			t.Errorf("want exactly %d", max)
+		}
+	}()
+
+	dialCtx, cancelDial := context.WithCancel(context.Background())
+	defer cancelDial()
+	dialer := &net.Dialer{}
+
+	var dialed, served int32
+	var pendingDials sync.WaitGroup
+	for n := attempts; n > 0; n-- {
+		wg.Add(1)
+		pendingDials.Add(1)
+		go func() {
+			defer wg.Done()
+
+			c, err := dialer.DialContext(dialCtx, l.Addr().Network(), l.Addr().String())
+			pendingDials.Done()
+			if err != nil {
+				t.Log(err)
+				return
+			}
+			atomic.AddInt32(&dialed, 1)
+			defer c.Close()
+
+			// The kernel may queue more than max connections (allowing their dials to
+			// succeed), but only max of them should actually be accepted by the
+			// server. We can distinguish the two based on whether the listener writes
+			// anything to the connection — a connection that was queued but not
+			// accepted will be closed without transferring any data.
+			if b, err := io.ReadAll(c); len(b) < len(msg) {
+				t.Log(err)
+				return
+			}
+			atomic.AddInt32(&served, 1)
+		}()
+	}
+
+	// Give the server a bit of time after it saturates to make sure it doesn't
+	// exceed its limit after serving this connection, then cancel the remaining
+	// dials (if any).
+	<-saturated
+	time.Sleep(10 * time.Millisecond)
+	cancelDial()
+	// Wait for the dials to complete to ensure that the port isn't reused before
+	// the dials are actually attempted.
+	pendingDials.Wait()
+	l.Close()
+	wg.Wait()
+
+	t.Logf("served %d simultaneous connections (of %d dialed, %d attempted)", served, dialed, attempts)
+
+	// If some other process (such as a port scan or another test) happens to dial
+	// the listener at the same time, the listener could end up burning its quota
+	// on that, resulting in fewer than max test connections being served.
+	// But the number served certainly cannot be greater.
+	if served > max {
+		t.Errorf("expected at most %d served", max)
+	}
+}
+
+func TestLimitListenerSaturation(t *testing.T) {
+	const (
+		max             = 5
+		attemptsPerWave = max * 2
+		waves           = 10
+		msg             = "bye\n"
+		timeout         = 1 * time.Second
+	)
+
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	l = ListenerWithTimeout(l, max, timeout)
+
+	acceptDone := make(chan struct{})
```

Comment:
why do we need this channel?

Reply:
This test comes from the library code. The usefulness of this channel is as a way to ensure all the go routines are terminated at the end of the test, essentially like an err group. I could have used an errgroup but kept with the semantics of the library tests.
When the test ends, the defer closes the listener then waits on the channel. The go routine has a `defer close(acceptDone)` that will unblock the defer if the Go routine itself finishes.